### PR TITLE
[ADD] website_sale_product_pack: New module

### PR DESCRIPTION
**Pull request by chienandalu**
Compatibility module between sale_product_pack and website_sale

cc @Tecnativa TT30385

### CODE REVIEW

File: website_sale_product_pack/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+from odoo import http
+
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+
+
+class WebsiteSaleProductPack(WebsiteSale):
+    @http.route()
+    def product(self, product, category="", search="", **kwargs):
+        res = super().product(product, category=category, search=search, **kwargs)
+        http.request.env.context = dict(http.request.env.context, whole_pack_price=True)
```
Comment (pedrobaeza) [inline website_sale_product_pack/controllers/main.py]: Please add docstring about this. I think there's a way of not doing Python hack (maybe through `qcontext`?)

File: website_sale_product_pack/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        """We need to keep the discount defined on the components when checking out"""
+        line = self.env["sale.order.line"].browse(line_id)
+        if line and line.pack_parent_line_id:
+            return super(
+                SaleOrder, self.with_context(pack_discount=line.discount)
+            )._cart_update(product_id, line_id, add_qty, set_qty, **kwargs)
+        return super()._cart_update(product_id, line_id, add_qty, set_qty, **kwargs)
+
+    def _website_product_id_change(self, order_id, product_id, qty=0):
+        """In the final checkout stop, we could miss the component discount as the
```
Comment (pedrobaeza) [inline website_sale_product_pack/models/sale_order.py]:
```suggestion
        """In the final checkout step, we could miss the component discount as the
```

File: website_sale_product_pack/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    def unlink(self):
+        """The website calls this method specifically. We want to get rid of
+        the children lines so the user doesn't have to"""
+        for line in self.filtered("pack_child_line_ids"):
+            self |= line.pack_child_line_ids
```
Comment (pedrobaeza) [inline website_sale_product_pack/models/sale_order.py]: Don't add over `self`, but have a new variable (`obj` for example), and call super with it. You can as well have it directly:
```suggestion
        obj = self + self.mapped("pack_child_line_ids")
```

File: website_sale_product_pack/models/sale_order.py (hunk 2 above)
Comment (chienandalu) [reply website_sale_product_pack/models/sale_order.py]: fixed. thanks :)

File: website_sale_product_pack/models/sale_order.py (hunk 3 above)
Comment (chienandalu) [reply website_sale_product_pack/models/sale_order.py]: fixed :+1:

File: website_sale_product_pack/controllers/main.py (hunk 1 above)
Comment (chienandalu) [reply website_sale_product_pack/controllers/main.py]: It didn't work as expected anyway. I'm removing it

File: website_sale_product_pack/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2021 Tecnativa - David Vidal
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Website Sale Product Pack",
+    "category": "E-Commerce",
+    "summary": "Compatibility module of product pack with e-commerce",
+    "version": "13.0.1.0.0",
+    "license": "AGPL-3",
+    "depends": ["website_sale", "sale_product_pack"],
```
Comment (pedrobaeza) [inline website_sale_product_pack/__manifest__.py:9]: Shouldn't this depend on `website_sale_stock`?

File: website_sale_product_pack/views/assets.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <template id="assets_frontend" inherit_id="website_sale.assets_frontend">
```
Comment (pedrobaeza) [inline website_sale_product_pack/views/assets.xml]: Isn't there specific assets for tests? Or at least load it on demo.

Comment (chienandalu) [reply website_sale_product_pack/views/assets.xml]: Yes. Done

File: website_sale_product_pack/__manifest__.py (hunk 4 above)
Comment (chienandalu) [reply website_sale_product_pack/__manifest__.py:9]: A pack of services would be valid as well

**Comment by chienandalu** [comment]
@ernestotejeda could you review?

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Ey, Travis goes to red now

**Comment by chienandalu** [comment]
:green_apple:

**Comment by nicomacr** [comment]
@chienandalu great work with this module. Really thanks
We got a similar module [here](https://github.com/ingadhoc/website/tree/13.0/website_product_pack) 

I've check this PR and compare with our module and find that would be interesting to have this features in the module to handle the publish mechanism:

- [here](https://github.com/ingadhoc/website/blob/13.0/website_product_pack/models/product_product.py#L13) we check for the products to add in a pack that all of them were published y if it not we alert. **Maybe with a domain in the field could be better.**

- [here](https://github.com/ingadhoc/website/blob/13.0/website_product_pack/models/product_template.py#L12) we set for the pack lines in a pack to be publish or not from the parent pack. And the inverse case if one of the pack component becomes to unpublish we set the same for the parent pack. **I think this part is better made in another method calling from write method.**
 
I thinks this could be necessary because the mixing of a product pack published with somes of the components not, could be confuse and some bug related to this will be happened.


Any thoughts about that?


Thanks again.

**Comment by chienandalu** [comment]
Thanks @nicomacr I propose you to add your review in order to merge this one, and then I'll gladly review your PR with those improvements :slightly_smiling_face:

**Review by nicomacr — APPROVED** [review]
LGTM

**Comment by pedrobaeza** [comment]
@chienandalu rebase to see the problem.

**Comment by alvaro-domatix** [comment]
We've detected a problem trying to use the packs on the website.

If you configure a product as a pack with pack_type = "detailed" and pack_component_price = "ignored" it apparently works while you do the cart until you try to pay. Once you enter the address information and click on "Next", the order sums the amount of all the components and the pack too.

Example:

Pack 1 (30€):
- Comp 1 (20€)
- Comp 2 (20€)
Once you click on next after the address information the order's amount is 30€ + 20€ + 20€ = 70€

File: website_sale_product_pack/readme/DESCRIPTION.rst [hunk 6]
```diff
@@ -0,0 +1,10 @@
+This module introduces compatibility of product packs with e-commerce:
+
+- In the cart summary, the components aren't editable and are shown hanging
+  from the main pack reference.
+- When we remove a pack from the cart, their components are removed as well.
+- The cart popup summary only shows the main pack line and discards the sublines in
+  the units count.
+- The cart summary shows the component lines hanging from the main one as well.
+- It's ensured the the prices are shown correctly for the whole pack and that they're
```
Comment (ernestotejeda) [inline website_sale_product_pack/readme/DESCRIPTION.rst:9]:
```suggestion
- It's ensured the prices are shown correctly for the whole pack and that they're
```

**Review by ernestotejeda — COMMENTED** [review]
I have functionally tested the module and I have found the same error that @alvaro-domatix  mentions here https://github.com/OCA/product-pack/pull/46#issuecomment-926829089
Also, when a pack is removed from the cart, its components are not removed.

**Pull request merged**